pkg/istio-agent/healthcheck.go

```diff
@@ -0,0 +1,198 @@
+	time.Sleep(w.config.InitialDelay)
+
+	numSuccess, numFail := 0, 0
+	if w.config.CheckType == HTTPHealthCheck {
+		for {
+			if code, err := httpCheck(w.config.HTTPConfig, w.config.ProbeTimeout); code >= 200 && code <= 299 {
+				numSuccess++
+				if numSuccess == w.config.SuccessThresh {
+					notifyHealthChange <- &discovery.DiscoveryRequest{TypeUrl: HealthInfoTypeURL}
+					numSuccess = 0
+				}
+			} else {
+				numFail++
+				if numFail == w.config.FailThresh {
+					notifyHealthChange <- &discovery.DiscoveryRequest{
+						TypeUrl: HealthInfoTypeURL,
+						ErrorDetail: &status.Status{
+							Code:    int32(code),
+							Message: err.Error(),
+						},
+					}
+					numFail = 0
+				}
+			}
+			// should this be a time.Ticker?
+			time.Sleep(w.config.CheckFrequency)
+		}
+	}
+
+	if w.config.CheckType == TCPHealthCheck {
+		for {
+			if err := tcpCheck(w.config.TCPConfig, w.config.ProbeTimeout); err == nil {
+				numSuccess++
+				if numSuccess == w.config.SuccessThresh {
+					notifyHealthChange <- &discovery.DiscoveryRequest{TypeUrl: HealthInfoTypeURL}
+					numSuccess = 0
+				}
+			} else {
+				numFail++
+				if numFail == w.config.FailThresh {
+					notifyHealthChange <- &discovery.DiscoveryRequest{
+						TypeUrl: HealthInfoTypeURL,
+						ErrorDetail: &status.Status{
+							Code:    int32(500),
+							Message: err.Error(),
+						},
+					}
+					numFail = 0
+				}
+			}
+		}
+	}
+
+	if w.config.CheckType == ExecHealthCheck {
+		for {
+			if err := execCheck(w.config.ExecConfig); err == nil {
+				numSuccess++
+				if numSuccess == w.config.SuccessThresh {
+					notifyHealthChange <- &discovery.DiscoveryRequest{TypeUrl: HealthInfoTypeURL}
+					numSuccess = 0
+				} else {
+					numFail++
+					if numFail == w.config.FailThresh {
+						notifyHealthChange <- &discovery.DiscoveryRequest{
+							TypeUrl: HealthInfoTypeURL,
+							ErrorDetail: &status.Status{
+								Code:    int32(500),
+								Message: err.Error(),
+							},
+						}
+						numFail = 0
+					}
+				}
+			}
+		}
+	}
+}
+
+// httpCheck performs a http get to a given endpoint with a timeout, and returns
+// the status and error.
+func httpCheck(config HTTPHealthCheckConfig, timeout time.Duration) (int, error) {
+	client := http.Client{
+		Timeout: timeout,
+	}
+	url := fmt.Sprintf("%s://localhost:%v/%s", config.Scheme, config.Port, config.Path)
+	resp, err := client.Get(url)
+	return resp.StatusCode, err
+}
+
+// tcpCheck connects to a given endpoint with a timeout, and errors
+// if connecting is unavailable (unhealthy)
+func tcpCheck(config TCPHealthCheckConfig, timeout time.Duration) error {
+	d := net.Dialer{
+		Timeout: timeout,
+	}
+	url := fmt.Sprintf("%s:%v", config.Host, config.Port)
+	conn, err := d.Dial("tcp", url)
+	if conn != nil {
+		conn.Close()
+	}
+	return err
+}
+
+// todo does adding Stderr to exec cmd return the error there?
+func execCheck(config ExecHealthCheckConfig) error {
+	healthCheckCmd := &exec.Cmd{
```

Comment:
Before we merge this we should get feedback from the product security WG @fpesce to ensure we aren't doing anything unsafe here. Executing arbitrary binaries needs some precautions